### ISSUE TITLE
Some Lessons Learned from Icon

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2931]
Some Lessons Learned from Icon<br>
<br>
https://ift.tt/2VPi01m<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<br>
<br>
<br>
<br>
<br>
<br>
via tratt.net https://tratt.net<br>
<br>
March 12, 2019 at 02:32PM
<table>
<tr>
<td valign="top" height="100%">
<div>
<div>
<table>
<tbody>
<tr>
<td>&nbsp;</td>
<td>
<h3>December 3 2007, last updated October 26 2010</h3>
<p><b>Updated (October 26 2007):</b> Before reading this article, please be aware that it has been superseded by the paper <a href="https://tratt.net/laurie/research/pubs/html/tratt__experiences_with_an_icon_like_expression_evaluation_system/">Experiences with an Icon-like expression evaluation system</a>.</p>
<p>One of <a href="http://convergepl.org/">Converge's</a> lesser known influences is <a href="http://www.cs.arizona.edu/icon/">Icon</a>. Although it's a relatively obscure language itself, a surprising number of people have heard of Icon's direct predecessor SNOBOL. I first heard of Icon through Tim Peters and Python, where Icon was the inspiration for Python's generator feature (basically procedures that whose execution can be resumed to produce multiple values). However Icon has a much richer, and definitely unique, approach to expression evaluation than any other imperative programming language. When I started working on Converge, I decided to go back to <q>the source</q>, and see how much of Icon's unique approach I could incorporate into Converge.</p>
<p>This article is a personal look back at Icon's influence on Converge, and the successes and failures resulting from that influence. It's quite probable that much of it reflects my slightly superficial understanding of Icon - apart from its innovative approach to expression evaluation, much of the language has an old fashioned feel to it, sometimes appearing like a dynamically typed version of Pascal, and this prevented me from ever wanting to write anything large in it. This isn't a criticism of Icon as such - it was ahead of its time in many ways - but is merely an observation that modern programmers expect something slightly different from their programming languages. Hopefully, even with the above caveat, this article contains something of value to those interested in programming languages.</p>
<h3>Why is Icon interesting</h3>
In a nutshell, Icon is interesting due to what it calls <em>goal-directed evaluation</em>. This is an evaluation mechanism which gives to an imperative programming language some of the flexibility more often associated with languages like Prolog, such as a limited form of backtracking. It is built around the concepts of success and failure, and generators (see the <a href="http://convergepl.org/documentation/current/quick_intro/">Converge documentation</a> for more details). Expressions can succeed or fail; if they succeed, they produce a value; if they fail, they do not produce a value and transmit failure to their containing scope. Generators are functions which can produce more than one value.
<p>Icon's syntax is derived from Algol, so the following complete program prints 1 to 9 inclusive:</p>
<pre>
procedure upto(x)
  i := 0
  while i &lt; x do {
    suspend i
    i := i + 1
  }
end

procedure main()
  every x := upto(10) do write(x)
end
</pre>
Most of this is probably reasonably intuitive, apart from the <code>every</code> and <code>suspend</code> keywords. <code>every</code> can be considered to be equivalent to the typical understanding of a <code>for</code> statement (indeed, in Converge, the same construct is called <code>for</code>). <code>suspend</code> is equivalent to <code>yield</code> in Python or Converge, returning a value, but allowing the procedure calls execution to be resumed directly after the <code>suspend</code> statement.
<h3>Failure and <code>if</code>: a beautiful combination</h3>
If I had to pick one thing from Icon that I would not be without in Converge, it would be the concept of failure in <code>if</code> statements. There is something beautiful about saying <q>if this function succeeded, assign the value to this variable and then do xyz</q>. This idiom is used to build a very useful convention that is present throughout Converge's libraries. As in most languages, container items such as dictionaries have a <code>get</code> function which returns the value associated with a certain key; <code>get</code>ing a key which doesn't exist generates an error. There is a very common variation on this use case, which is to check whether a dictionary has a certain key, and if it does to do something with its value; otherwise a different code path is taken. In most languages this idiom is expressed roughly as follows:
<pre>
d := Dict{"a" : 2, "b" : 8}
if d.has_key("a"):
  Sys::println(d["a"])
</pre>
Not only is the double lookup of <code>a</code> an eyesore, and a maintenance accident waiting to happen, but it can be a significant overhead in tight loops. In Python (and probably other languages), it's common to see the following idiom:
<pre>
d := {"a" : 2, "b" : 8}
try:
  v = d.has_key("a")
  print v
except KeyError:
  pass
</pre>
This idiom makes use of the fact that it's nearly always quicker to catch an exception if a key isn't found than to perform two lookups. In my opinion, this idiom is far from pleasant, for reasons that I'm sure most readers will intuitively understand.
<p>In Converge one can express this common idiom thus:</p>
<pre>
x := Dict{"a" : 2, "b" : 8}
if v := x.find("a"):
  Sys::println(v)
</pre>
In this example, <code>v</code> only gets assigned a value if <code>find</code> succeeds. There are some other advantages to this approach (e.g. there's no fiddling around checking for <code>null</code>), but the symmetry between <code>get</code> and <code>find</code>, and the consistency of this convention across libraries has proved a real success in Converge. For me, this feature alone justifies the <q>Icon experiment</q> in Converge.
<h3>Variables should not spring into existence with a default value</h3>
The idiom mentioned in the previous section is, unfortunately, somewhat dangerous in Icon itself, as all variables have a default value of sorts. What this means is that the following Icon code runs without error:
<pre>
procedure main()
  if 1 &lt; 0 then x := 0
  write(x)
end
</pre>
To me, this is odd, because <code>x</code> never has a value assigned to it (as the expression in the <code>if</code> statement is clearly false); any errors resulting from such a mistake turn out to be very difficult to debug. In Converge I therefore made it so that reading from a currently unassigned variable raises an error, which protects against programming slips such as the following:
<pre>
x := Dict{"a" : 2, "b" : 8}
if v := x.find("c"):
  ...
else:
  Sys::println(v)
</pre>
As the lookup of <q>c</q> fails, <code>v</code> doesn't have a value assigned to it, so reading from <code>v</code> in the <code>else</code> branch of the <code>if</code> statement raises an <code>Unassigned_Var_Exception</code>, pinpointing the offending code.
<h3>Procedures shouldn't fail by default</h3>
In Icon, the default <q>return value</q> of a procedure is failure. This makes quite a lot of sense for generators such as the following:
<pre>
procedure upto(x)
  i := 0
  while i &lt; x do {
    suspend i
    i := i + 1
  }
end
</pre>
As this suggests, the typical action for a generator is (via a loop) to generate several values; once that loop is finished, the generator fails. Thus each Icon procedure effectively has <code>return &amp;fail</code> at its end. Very early versions of Converge inherited this behaviour.
<p>What became quickly apparent is that while this is reasonable behaviour for generators, it can be a disaster for normal procedures. Look at this code and see if you can work out what will happen when it's executed:</p>
<pre>
procedure f(x)
  if x &gt; 0 then return 1
end

procedure main()
  write(f(-1))
end
</pre>
If you guessed that nothing would be printed to screen, congratulations. If you didn't, then don't worry, because this frequently surprised me. This happens since the failure of the <code>if</code>'s condition means that the procedure executes its default return action, which is to fail. Thus the procedure doesn't return a value, and the call to <code>write</code> is never even evaluated.
<p>Although this might seem trivial, the problem becomes significantly worse when it's embedded in a large body of code. <code>f</code> is an example of what I informally think of as the <q>dangling <code>if</code> return problem</q> - in other words, it's quite easy to write a procedure where different values should be returned, but where one branch of an <code>if</code> incorrectly neglects to actually return anything. I find this occurs quite often during the early stages of development when one is fleshing out functions. On several occasions I spent several hours tracking down instances of this problem.</p>
<p>The question I asked myself was whether this feature was worth the pain. A quick analysis of real world code quickly showed that the vast majority of functions aren't generators, and indeed only ever return a single value. Therefore optimising for the exceptional case (generators) is an odd design choice.</p>
<p>I considered two solutions to the problem. First, one could syntactically differentiate generators and procedures, with the former failing by default and the latter not. Second, one can make all procedures not fail by default. Adding new syntax is a decision that should never be taken lightly, and since generators aren't that common, I couldn't justify the added conceptual complexity. Therefore Converge functions became similar to Python functions, with their default return action being to return the <code>null</code> object.</p>
<h3>Generators are useful, but easily hidden</h3>
Generators are an integral part of Icon and Converge, but as well as having no specific syntax to define them, there is no specific syntax to call them. Assuming that <code>g</code> is a generator, I have grown quite fond of the following idiom:
<pre>
l := []
...
for l.append(g())
</pre>
which appends every element generated by <code>g</code> to <code>l</code>. However the problem in such cases is that it can be difficult to work out what part of the expression is the generator. Converge now tries to solve this problem by prefixing generator function names, whenever it makes sense, with <code>iter_</code> which highlights that the function in question is a generator. This makes it easier to look at a piece of code and work out what it does without having to understand every function called.
<h3>Backtracking is rarely useful</h3>
One of the major features of goal-directed evaluation is that backtracking can be performed. The most common way for this to be achieved is by linking expressions together with <code>&amp;</code>. The resulting <em>conjunction</em> of these expressions only succeeds if all of its component expressions succeed. The expressions are evaluated in order. If one of them fails, and a previous expression is a generator, then backtracking occurs. The previous generator is resumed to produce a new value, and the conjunction then resumes execution from that point.
<p>Combining conjunction with <code>for</code> allows compact expressions such as the following to be expressed:</p>
<pre>
x := "abcabaacvcabcbab"
for Sys::println(i := 0.iter_to(x.len()) &amp; i % 2 == 0 &amp; x[i] == "a" &amp; i)
</pre>
This prints out every index position in <code>a</code> which is a multiple of two, and which contains the character <code>a</code> (<code>0 6 10 14</code> for those who are interested).
<p>There are two problems with the above. Most obviously, it is incredibly difficult to read from a human perspective; indeed, it would be far better written as a couple of <code>if</code> statements within a <code>for</code> body. The second problem is that the form of backtracking used is often too weak to be useful. Icon allows variables in conjunctions to have assignments undone during backtracking, but even this isn't really enough (I didn't even bother implementing such functionality in Converge, because I couldn't really see its use). Full backtracking often involves undoing assignments within objects and so on, and no imperative language can ever hope to do this automatically.</p>
<p>Out of all the systems I've written in Converge, only one has made any practical use of the built-in backtracking, and even then it needed some manual help to be truly useful.</p>
<h3>The <code>fail</code> variable causes bizarre behaviour</h3>
One way in which Icon shows its age is its differentiation between values and references. In keeping with most object orientated languages, Converge makes no such distinction to the user (although within the VM it does so for efficiencies sake). One of Icon's main oddities is that <code>fail</code> is a synonym for <code>return &amp;fail;</code>, while <code>&amp;fail</code> is a sort-of reference to an invisible object.
<p>Converge simplified this, so that <code>fail</code> was a variable pointing to a semi-special object (in similar fashion to the <code>null</code> object). However the <code>fail</code> variable proved, in admittedly rare cases, to be conceptually troubling. It's troubling in Icon too, although in slightly different ways. What do you think happens in Icon if I define <code>l</code> as the following and then iterate over each of its elements and print them out?</p>
<pre>
l := [1, &amp;fail;, 3]
</pre>
If you guessed that an error is raised because <code>l</code> didn't get assigned a value (because evaluating <code>&amp;fail</code> cause the whole thing to fail), pat yourself on the back. Now try this:
<pre>
l := []
put(l, 1)
put(l, &amp;fail;)
put(l, 3)
</pre>
If you guessed that <code>1</code>, then <code>2</code> get printed out, you're doing very well. And finally consider this:
<pre>
l := []
put(l, 1)
t := &amp;fail;
put(l, t)
put(l, 3)
</pre>
If you guessed that <code>1</code>, then a blank line, and then <code>2</code> get printed out, you're doing much better than I ever managed to.
<p>Converge's attempts to simplify thus were somewhat successful, but left one hole. While one could successfully evaluate the following list in Converge:</p>
<pre>
l := [1, fail, 2]
</pre>
the following mysteriously printed nothing:
<pre>
Sys::println(l[1])
</pre>
since <code>l[1]</code> is a synonym for <code>l.get(1)</code>, which of course tried to evaluate <code>return fail</code>, which then caused <code>get</code> to fail. While this might seem a contrived example, it unavoidably cropped up when getting a module to return the value of a specific definition, since one of those definitions was <code>fail</code>.
<p><code>fail</code> was edged out of Converge in stages. A while ago, the recommendation became that the only safe idiom involving <code>fail</code> was <code>return fail</code>. Perhaps surprisingly, this was not an onerous restriction. Nevertheless it still failed to prevent the problem of <code>fail</code> being a definition in a module. Therefore the <code>fail</code> variable was finally banished entirely from Converge recently (although it lives on internally in the VM). <code>fail</code> is now an expression in Converge's grammar, and is equivalent to <code>return &amp;fail</code> in Icon. Thus finally, all of the bizarre behaviour associated with the <code>fail</code> variable is forgotten, and I now sleep easier at night.</p>
<h3>Do something different</h3>
When I was looking for information in Icon, I stumbled across a few interviews with Ralph Griswold (Icon's main designer), and a paper on its history. One thing that became clear to me was that a major design goal of Icon was not to be just another (boring) synthesis of a few existing language features - like the vast majority of programming languages - but to try out genuinely new things. Have a look at e.g. p38 of <a href="http://www.cbi.umn.edu/oh/pdf.phtml?id=135">this interview</a> or p6 and 13 of <a href="http://dl.acm.org/citation.cfm?id=1057830">this paper looking back at Icon's history</a>. Whatever you might think of Icon, it undoubtedly fulfilled this aim: its expression evaluation system alone has no precedent in contemporary programming languages (and there are other interesting parts of the language).
<p>When I was starting Converge, I agreed whole heartedly with this aim, but had no idea how to achieve it. Indeed, when Converge started, I only had the vague intention of chucking a few seemingly useful influences into the melting pot (initially Python, Icon, ObjVLisp, and ultimately Template Haskell when I'd got the mix of the former three languages about right). I had that vague feeling that either everything interesting had been done before, or that the remaining interesting things were outside of my grasp. To be honest, I sort of gave up on the aim of being innovative, and just concentrated on trying to mix Converge's influences together in a coherent way. What has been interesting is that as Converge has developed, new challenges have presented themselves, and I've had to find answers (some better than others) to such challenges. And in so doing, I think Converge has grown some genuinely innovative features, such as <a href="https://tratt.net/laurie/blog/entries/filling_in_a_gap">DSL blocks, and its approach to error information for macros</a> amongst others. And in Converge's case - unlike Icon, as I understand it - the journey isn't over. Indeed, since I think Converge is currently nearing the end of the beginning, there's bound to be new challenges ahead, some of which will lead to new innovations of one sort or another, some ultimately successful, some ultimately not.</p>
<p>When I read about Icon's aim, and looked at the end result, I assumed that the innovative ideas in Icon had sprung out of nowhere. I, on the other hand, had no idea how to bring such ideas, fully formed, into the world. What's become clear to me is a restating of Edison's famous rule. Innovation is often a matter of perspiration over inspiration. Now my advice to those who wish to do innovative work is simple: if you put the miles in, you'll raise your head one day and discover that you're in a place where no-one else has ever found themselves in before. It's a good feeling, and one that benefits us all, as collectively we gradually discover what works and what doesn't.</p>
<p><b>Updated (December 10 2007):</b> Used a clearer example in the <q>Failure and if: a beautiful combination</q> section.</p>
<p><b>Updated (October 26 2007):</b> This article has been superseded by the paper <a href="https://tratt.net/laurie/research/pubs/">Experiences with an Icon-like expression evaluation system</a>.</p>
<p><a href="https://twitter.com/laurencetratt">Follow me on Twitter</a></p>
<p><a href="https://twitter.com/laurencetratt">Follow me on Twitter</a></p>
</td>
<td>&nbsp;</td>
<td></td>
</tr>
</tbody>
</table>
</div>
</div>
</td>
</tr>
</table>
</div>
</div>
